### PR TITLE
HHH-4808 - SessionImpl.initializeCollection() does not release JDBC connection (if outside of a transaction)

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-4808

### CODE REVIEW

**Comment by gsmet** [comment]
Haven’t reviewed the PR but the JIRA number of the second commit is incorrect :)

**Comment by dreab8** [comment]
fixed
Thanks @gsmet

File: hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java [hunk 1]
```diff
@@ -2241,11 +2248,18 @@ public SessionFactoryImplementor getSessionFactory() {
...
+		try {
+			InitializeCollectionEvent event = new InitializeCollectionEvent( collection, this );
+			for ( InitializeCollectionEventListener listener : listeners( EventType.INIT_COLLECTION ) ) {
+				listener.onInitializeCollection( event );
+			}
+			success = true;
+		}
+		finally {
+			delayedAfterCompletion();
+			afterOperation( success );
```
Comment (gsmet) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java]: When initializing an entity and its collections in one go, we will now call `afterOperation()` several times (after each collection initialization). It won't have any consequence (e.g. it won't close the connection or close the transaction and reopen a new one for the next collection)?

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java]: I think what could happen is that after the connection is closed any other operation will cause an exception due to the closed connection. 
I'm going to create a test for the scenario you proposed.

**Review by sebersole — CHANGES_REQUESTED** [review]
To me the proper way to do this would be in `AbstractPersistentCollection#withTemporarySessionIfNeeded`, no?

**Comment by dreab8** [comment]
superseded by https://github.com/hibernate/hibernate-orm/pull/3868

**Pull request closed without merging**